### PR TITLE
py: Support object __getattr__, __delattr__ and __setattr__ methods

### PR DESCRIPTION
**Pull request by dmazzella**
Related to #2375

### CODE REVIEW

File: py/objobject.c [hunk 1]
```diff
@@ -59,13 +59,48 @@ STATIC mp_obj_t object___new__(mp_obj_t cls) {
...
+#endif
+
 STATIC const mp_rom_map_elem_t object_locals_dict_table[] = {
     #if MICROPY_CPYTHON_COMPAT
     { MP_ROM_QSTR(MP_QSTR___init__), MP_ROM_PTR(&object___init___obj) },
     #endif
     #if MICROPY_CPYTHON_COMPAT
     { MP_ROM_QSTR(MP_QSTR___new__), MP_ROM_PTR(&object___new___obj) },
     #endif
+    #if (MICROPY_CPYTHON_COMPAT && MICROPY_PY_ATTRS_METHODS)
```
Comment (stinos) [inline py/objobject.c]: Why not just #if MICROPY_PY_ATTRS_METHODS ?

File: py/mpconfig.h [hunk 2]
```diff
@@ -636,6 +636,12 @@ typedef double mp_float_t;
 #define MICROPY_PY_DESCRIPTORS (0)
 #endif
 
+// Whether to support object (__getattr__, __delattr__ and __setattr__)
+// This costs some code size and makes all load attrs and store attrs slow
```
Comment (stinos) [inline py/mpconfig.h]: 'slow*er*' would be better wording imo, without measuring the actual impact it's hard to make absolute statements about it.

**Comment by dmazzella** [comment]
Is correct to use: `mp_obj_dict_get_map(self_in)`

instead of:
```
mp_obj_type_t *type = mp_obj_get_type(self_in);
if (type->locals_dict != NULL) {
    mp_map_t *locals_map = &type->locals_dict->map;
   ...
```

??

File: py/objobject.c (hunk 1 above)
Comment (stinos) [inline py/objobject.c]: Ah didin't see that in the github diff. Still it's a bit confusing as now one might use `#define MICROPY_PY_ATTRS_METHODS 1` but it won't have any effect unless MICROPY_CPYTHON_COMPAT is also defined. But maybe there's no other way.

**Comment by stinos** [comment]
Doesn't seem equivalent at first sight. Does everything still work if you use `#define MICROPY_BUILTIN_METHOD_CHECK_SELF_ARG 0`? That would make the `mp_check_self` macro have effect so mp_obj_dict_get_map will assert unless a dict is passed in.

**Comment by dmazzella** [comment]
The pull request is not ready ...

```
$ ./micropython.exe  ../tests/basics/getattr.py
Traceback (most recent call last):
  File "../tests/basics/getattr.py", line 10, in <module>
  File "../tests/basics/getattr.py", line 5, in __init__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
  File "../tests/basics/getattr.py", line 8, in __getattr__
RuntimeError: maximum recursion depth exceeded
```

**Comment by dpgeorge** [comment]
To enable the tests, please enable this new feature in unix/mpconfigport_coverage.h and then also add a new test file.

**Comment by dmazzella** [comment]
using `super(A, self).__setattr__('d', d)` instead of `self.d = d` work

suggest?

**Comment by dmazzella** [comment]
@dpgeorge  Solved RuntimeError: maximum recursion depth exceeded

**Comment by dmazzella** [comment]
Rebase

**Pull request closed without merging**